src/c/ubench.c

```diff
@@ -16,44 +16,139 @@
+	// loaded as an agent, so this is just an extra precaution.
+	//
+	static ubench_atomic_int_t startup_guard = { .atomic_value = 0 };
+
+	bool initialized = ubench_atomic_int_inc(&startup_guard) != 0;
+	if (initialized) {
+		DEBUG_PRINTF("ubench already initialized.");
+		return true;
+	}
+
+	//
+	// Initialize the individual modules. Some of them are allowed to fail,
+	// because they will try to use JVMTI which may be unavailable (e.g., in
+	// the native image). This will make some functionality unavailable, but
+	// will still allow using the agent e.g., to use PAPI.
+	//
+	DEBUG_PRINTF("initializing counters module");
+	if (!ubench_counters_init(jvm)) {
+		WARN_PRINTF("failed to initialize event counter module.");
 	}
 
-	rc = ubench_counters_init(jvmti);
+	DEBUG_PRINTF("initializing threads module");
+	if (!ubench_threads_init(jvm)) {
+		WARN_PRINTF("failed to initialize threads module.");
+	}
 
-	if (rc != JNI_OK) {
-		return JNI_ERR;
+	DEBUG_PRINTF("initializing measurement module");
+	if (!ubench_measurement_init()) {
+		ERROR_PRINTF("failed to initialize measurement module.");
+		return false;
 	}
 
-	rc = ubench_benchmark_init();
-	if (rc != JNI_OK) {
-		return JNI_ERR;
+	DEBUG_PRINTF("initializing event module");
+	if (!ubench_event_init()) {
+		ERROR_PRINTF("failed to initialize event module.");
+		return false;
 	}
 
-	DEBUG_PRINTF("initialization completed successfully.");
-	return JNI_OK;
+	return true;
+}
+
+static void
+ubench_shutdown(JavaVM* UNUSED_PARAMETER(jvm)) {
+	assert(jvm != NULL);
+
+	// Avoid finalizing the library twice on shutdown.
+	static ubench_atomic_int_t shutdown_guard = { .atomic_value = 0 };
+
+	bool finalized = ubench_atomic_int_inc(&shutdown_guard) != 0;
+	if (finalized) {
+		DEBUG_PRINTF("ubench already finalized.");
+		return;
+	}
+
+	// So far, there is nothing to do on shutdown.
+	// We do not dispose the JVMTI environments.
+}
+
+//
+
+/**
+ * Dummy function used by the 'UbenchAgent' class to check if the native
+ * functions are already bound. If the call to this function succeeds, it
+ * means that the agent has been already loaded and there is no need to
+ * try loading it via the 'loadLibrary()' method which could fail if the
+ * agent was loaded using the '-agentlib' JVM option without setting the
+ * value of the 'java.library.path'.
+ */
+JNIEXPORT jboolean JNICALL
+Java_cz_cuni_mff_d3s_perf_UbenchAgent_nativeIsLoaded(
+	JNIEnv* UNUSED_PARAMETER(jni), jclass UNUSED_PARAMETER(ubench_class)
+) {
+	DEBUG_PRINTF("nativeIsLoaded called from Java");
+	return JNI_TRUE;
+}
+
+//
+
+JNIEXPORT jint JNICALL
+Agent_OnLoad(JavaVM* vm, char* UNUSED_PARAMETER(options), void* UNUSED_PARAMETER(reserved)) {
+	DEBUG_PRINTF("agent loading started.");
+	bool success = ubench_startup(vm);
+	DEBUG_PRINTF("agent loading finished (%s).", success ? "success" : "failure");
+	return success ? JNI_OK : JNI_ERR;
+}
+
+JNIEXPORT void JNICALL
+Agent_OnUnload(JavaVM* vm) {
```

Comment:
Good point, I was thinking about it while I was away. We could either run test using both methods, or run a just single test using the other method. Any preferences?

Reply:
Any preferences on how would you like it to be integrated?

- One option is to modify the `test` target in `build.xml` so that it would use different loading method for some of the tests (I could not attach a diff, so I'm including it inline). Alternatively, I could modify the `list-events` (or `demo`) target to use `-Djava.library.path` instead of `-agentpath`. All those targets are used in the `gate` target, so modifying any of them should do.

```diff
--- a/build.xml
+++ b/build.xml
@@ -685,53 +685,55 @@
 				<pathelement path="${hamcrest.jar.path}" />
 			</classpath>
 			<compilerarg value="-Xlint:all"/>
 		</javac>
 	</target>
 
 	<macrodef name="self-test">
 		<attribute name="classname" default="mixed" />
+		<attribute name="loadarg" default="-agentpath:${agent.path}" />
 		<attribute name="jvmarg" default="-Xmixed" />
 		<attribute name="progarg" default="mixed" />
 
 		<sequential>
-			<echo level="info" message="[run] java -agentpath:${agent.path} @{jvmarg} @{classname} @{progarg}" />
+			<echo level="info" message="[run] java @{loadarg} @{jvmarg} @{classname} @{progarg}" />
 			<java
 				classname="@{classname}"
 				fork="true"
 				failonerror="true"
 			>
 				<classpath>
 					<pathelement path="${classes.build.dir}" />
 					<pathelement path="${test.classes.build.dir}" />
 				</classpath>
-				<jvmarg value="-agentpath:${agent.path}" />
+				<jvmarg value="@{loadarg}" />
 				<jvmarg value="@{jvmarg}" />
 				<arg value="@{progarg}" />
 			</java>
 		</sequential>
 	</macrodef>
 
 	<target name="test"
 		depends="compile,compile-test"
 		description="Run simple self-tests (obsoleted)."
 	>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.CompilationCounterTest"
 			jvmarg="-Xmixed"
 			progarg="mixed"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.CompilationCounterTest"
 			jvmarg="-Xint"
 			progarg="int"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.BenchmarkTest"
+			loadarg="-Djava.library.path=${agent.build.dir}"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.ThreadTimeTest"
 			progarg="4"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.MeasurementSamplingTest"
 		/>
@@ -808,17 +810,17 @@
 			classname="cz.cuni.mff.d3s.perf.demo.ListEvents"
 			fork="true"
 			failonerror="true"
 		>
 			<classpath>
 				<pathelement path="${classes.build.dir}" />
 				<pathelement path="${demo.classes.build.dir}" />
 			</classpath>
-			<jvmarg value="-agentpath:${agent.path}" />
+			<jvmarg value="-Djava.library.path=${agent.build.dir}" />
 		</java>
 	</target>
 
 	<target name="dist"
 		description="Create distributable ZIP with all sources."
 	>
 		<mkdir dir="${build.dir}" />
 		<zip destfile="${build.dir}/${dist.name}.zip">
```

- Another option is to modify the `self-test` macro to run each test twice, one as agent and one as native library. This would avoid duplicating the information from the `test` target, but the `test` target would run twice as long.
- Finally, I could create a separate target, e.g., `test-loading`, that would test different methods on one test, e.g., which will result in some duplication of entries found in the `test` or `list-events` or `demo` targets.

Reply:
Another option: use the agent in `test` target and the `loadLibrary()` approach in `test-junit` target.

This required a fix both in the `NativeThreads` tests and the threads module initialization, so it actually flushed out some problems, which is good. The change in `build.xml` is then minimal.

Reply:
OK, now I just need to find out why it fails on MacOS. A platform that causes even more problems than Windows should be banned.

Reply:
I would run it both styles in `gate` (i.e. run both JUnit tests and self-tests in both versions). I hope that we can pass some extra attribute to the target to specify this.

Reply:
> I would run it both styles in gate (i.e. run both JUnit tests and self-tests in both versions). I hope that we can pass some extra attribute to the target to specify this.

Done. All targets that use the agent are now run both ways in the `gate` target.

Reply:
Thanks!